src/Workspaces/Core/Portable/Shared/Utilities/NameGenerator.cs

```diff
@@ -1,91 +1,121 @@
+
+            var result = ArrayBuilder<string>.GetInstance(names.Length);
+            result.AddRange(names);
+            EnsureUniquenessInPlace(result, isFixed, canUse, isCaseSensitive);
+            return result.ToImmutableAndFree();
         }
 
         /// <summary>
-        /// Transforms baseName into a name that does not conflict with any name in 'reservedNames'
+        /// Ensures that any 'names' is unique and does not collide with any other name.  Names that
+        /// are marked as IsFixed can not be touched.  This does mean that if there are two names
+        /// that are the same, and both are fixed that you will end up with non-unique names at the
+        /// end.
         /// </summary>
-        public static string EnsureUniqueness(
-            string baseName,
-            IEnumerable<string> reservedNames,
+        public static ImmutableArray<string> EnsureUniqueness(
+            ImmutableArray<string> names,
+            ImmutableArray<bool> isFixed,
+            Func<string, bool>? canUse = null,
             bool isCaseSensitive = true)
         {
-            var names = new List<string> { baseName };
-            var isFixed = new List<bool> { false };
+            using var isFixedDisposer = ArrayBuilder<bool>.GetInstance(names.Length, out var isFixedBuilder);
+            isFixedBuilder.AddRange(isFixed);
 
-            names.AddRange(reservedNames.Distinct());
-            isFixed.AddRange(Enumerable.Repeat(true, names.Count - 1));
+            var result = ArrayBuilder<string>.GetInstance(names.Length);
+            result.AddRange(names);
 
-            var result = EnsureUniqueness(names, isFixed, isCaseSensitive: isCaseSensitive);
-            return result.First();
+            EnsureUniquenessInPlace(result, isFixedBuilder, canUse, isCaseSensitive);
+
+            return result.ToImmutableAndFree();
         }
 
-        private static void EnsureUniquenessInPlace(
-            IList<string> names,
-            IList<bool> isFixed,
-            Func<string, bool> canUse,
+        /// <summary>
+        /// Updates the names in <paramref name="names"/> to be unique.  A name at a particular
+        /// index <c>i</c> will not be touched if <c>isFixed[i]</c> is <see langword="true"/>. All
+        /// other names will not collide with any other in <paramref name="names"/> and will all
+        /// return <see langword="true"/> for <c>canUse(name)</c>.
+        /// </summary>
+        public static void EnsureUniquenessInPlace(
+            ArrayBuilder<string> names,
+            ArrayBuilder<bool> isFixed,
+            Func<string, bool>? canUse = null,
             bool isCaseSensitive = true)
         {
-            canUse ??= (s => true);
+            canUse ??= Functions<string>.True;
+
+            using var disposer = ArrayBuilder<int>.GetInstance(out var collisionIndices);
```

Comment:
this is really the most important part.  instead of allocating this collection each time through the loop, we can simply make a single instance we reuse.